Dalvik Bytecode deocding

gnieto opened the pull request:
I've implemented most of the Dalvik operations (there are some instructions missing: _binop/lit16_, _ binop/lit8_, _invoke-polymorphic_, _invoke-polymorphic/range_, _invoke-custom_, _invoke-custom/range_).

I've created a `ByteCodeDecoder` which is an iterator that yeilds `ByteCode` from the source.

This is not used anywhere else (yet), as the code segments are not yet parsed on develop.

gnieto committed:
Struct

gnieto committed:
Some bytecodes

gnieto committed:
More move operations

gnieto committed:
Move object bytecodes

gnieto committed:
Move results bytecodes

gnieto committed:
Return bytecods

gnieto committed:
Add constant byte codes

gnieto committed:
Fix 12x decoder order

gnieto committed:
Decode const16

gnieto committed:
Bit extend to i32

gnieto committed:
Const High16

gnieto committed:
ConstWide16

gnieto committed:
ConstWide32

gnieto committed:
Add ConstWide

gnieto committed:
Add ConstWideHigh16

gnieto committed:
Add ConstString

gnieto committed:
Add ConstStringJumbo

gnieto committed:
Add ConstClass

gnieto committed:
Add Monitor bytecodes

gnieto committed:
Add CheckCast

gnieto committed:
Add InstanceOf

gnieto committed:
Add ArrayLength

gnieto committed:
Add NewInstance

gnieto committed:
Add NewArray

gnieto committed:
Add FilledNewArray

gnieto committed:
Add ArrayFillRange

gnieto committed:
Add FillArrayData

gnieto committed:
Add Throw

gnieto committed:
Add Goto

gnieto committed:
Add Goto16

gnieto committed:
Add Goto32

gnieto committed:
Add PackedSwitch

gnieto committed:
Some printing formatting

gnieto committed:
Add SparseSwitch

gnieto committed:
Add Cmp

gnieto committed:
Add If bytecode

gnieto committed:
Add If0

gnieto committed:
Add ArrayOperations

gnieto committed:
Added instance and static operations

gnieto committed:
Add Invoke operations

gnieto committed:
Add Unary operations

gnieto committed:
Add Binary operations

Razican reviewed (changes requested):
Incredible work!! Thanks!

Seems that tests are failing due to formatting. Could you check it? Also, seems we have a lot of warnings from missing docs. Maybe it's not so important right now, but it will be in the future. Let me know what you think.

gnieto committed:
Add BinaryLit8 operations

gnieto committed:
Apply rustfmt

gnieto committed:
Clean unneded uses and add some documentation blocks

gnieto commented:
I've added the allow missing docs on the bytecodes structures/enums as it will not add any value at the moment (if you are using this, probably you are also looking into dalvik documentation, as it will be a nicer source of data).

Besides this, formatting is ok now and I've added some new documentation.

Razican commented on src/lib.rs:
```diff
@@ -103,7 +104,7 @@ impl Dex {
 
 impl From<DexReader> for Dex {
-    fn from(reader: DexReader) -> Dex {
+    fn from(_: DexReader) -> Dex {
```
Actually, I like this warning, since it remembers what it's still not implemented.

gnieto committed:
Restore input parameter name

gnieto committed:
Add InvokePolymorphic

gnieto committed:
Add InvokePolymorphicRange

gnieto committed:
Add InvokeCustom

gnieto committed:
Add InvokeCustomRange

gnieto commented:
Added missing instructions

gnieto committed:
Fix some clippy warnings

Razican commented:
Amazing work!!! 🎉
